codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/XmlShapeSerVisitor.java

```diff
@@ -0,0 +1,259 @@
+            MemberShape valueMember = shape.getValue();
+            Shape valueTarget = model.expectShape(valueMember.getTarget());
+            String valueName = valueMember.getTrait(XmlNameTrait.class)
+                    .map(XmlNameTrait::getValue)
+                    .orElse("key");
+            writer.write("const keyNode = new __XmlNode($S');", valueName);
+            writer.write("keyNode.addChildNode($L)", valueTarget.accept(getMemberVisitor("input[key]")));
+            writer.write("entryNode.addChildNode(keyNode);");
+
+            // Add the entry to the collection.
+            writer.write("collectedNodes.push(entryNode);");
+        });
+
+        writer.write("return collectedNodes;");
+    }
+
+    @Override
+    protected void serializeStructure(GenerationContext context, StructureShape shape) {
+        TypeScriptWriter writer = context.getWriter();
+        writer.addImport("XmlNode", "__XmlNode", "@aws-sdk/xml-builder");
+
+        // Handle the @xmlName trait for the structure itself.
+        String nodeName = shape.getTrait(XmlNameTrait.class)
+                .map(XmlNameTrait::getValue)
+                .orElse(shape.getId().getName());
+
+        // Create the structure's node.
+        writer.write("const bodyNode = new __XmlNode($S);", nodeName);
+
+        // Add @xmlNamespace value of the structure to the node.
+        AwsProtocolUtils.writeXmlNamespace(context, shape, "bodyNode");
+
+        // Serialize every member of the structure if present.
+        Map<String, MemberShape> members = new TreeMap<>(shape.getAllMembers());
+        members.forEach((memberName, memberShape) -> {
+            String inputLocation = "input." + memberName;
+            writer.openBlock("if ($L !== undefined) {", "}", inputLocation, () -> {
+                serializeNamedMember(context, memberName, memberShape, () -> inputLocation);
+            });
+        });
+
+        writer.write("return bodyNode;");
+    }
+
+    void serializeNamedMember(
+            GenerationContext context,
+            String memberName,
+            MemberShape memberShape,
+            Supplier<String> inputLocation
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        DocumentMemberSerVisitor inputVisitor = getMemberVisitor(inputLocation.get());
+
+        // Use the @xmlName trait if present on the member, otherwise use the member name.
+        String locationName = memberShape.getTrait(XmlNameTrait.class)
+                .map(XmlNameTrait::getValue)
+                .orElse(memberName);
+        // Grab the target shape so we can use a member serializer on it.
+        Shape target = context.getModel().expectShape(memberShape.getTarget());
+
+        // Handle @xmlAttribute simple members.
+        if (memberShape.hasTrait(XmlAttributeTrait.class)) {
+            writer.write("bodyNode.addAttribute($S, $L);", locationName, inputLocation.get());
+        } else {
```

Comment:
This seems like it should be broken into multiple methods with really readable method names.